x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -261,13 +267,33 @@ public void restoreShard(IndexShard indexShard, SnapshotId snapshotId, Version v
         String nodeId = response.getNodeId();
         // TODO: Implement file restore
         closeSession(remoteClient, nodeId, sessionUUID);
+        maybeUpdateMappings(client, remoteClient, leaderIndex, indexShard.indexSettings());
     }
 
     @Override
     public IndexShardSnapshotStatus getShardSnapshotStatus(SnapshotId snapshotId, Version version, IndexId indexId, ShardId leaderShardId) {
         throw new UnsupportedOperationException("Unsupported for repository of type: " + TYPE);
     }
 
+    private void maybeUpdateMappings(Client localClient, Client remoteClient, Index leaderIndex, IndexSettings followerIndexSettings) {
```

Comment:
@martijnvg would it make sense to share this code with the one in ShardFollowTasksExecutor?

Reply:
Conceptually this code is doing the same as the update mapping code in shard follow task, but the update mapping code is tightly coupled with `ShardFollowTasksExecutor` and `ShardFollowNodeTask`. Also the mapping update code is asynchronous over there and here it is synchronous.

What I think is possible, is that factory methods for both `ClusterStateRequest` and `PutMappingRequest` are introduced that is then used here and in `ShardFollowTasksExecutor`.